design/one-pager-build-with-earthly.md

```diff
@@ -0,0 +1,249 @@
+
+## Proposal
+
+I proposed we switch from Make to https://earthly.dev.
+
+Earthly targets the 'glue' layer between language-specific tools like `go` and
+CI systems like GitHub Actions. In Crossplane, Earthly would replace Make and
+Docker. It's based on Docker's [BuildKit][buildkit], so all builds are
+containerized and hermetic.
+
+### Configuration
+
+The Earthly equivalent of a `Makefile` is an `Earthfile`. An `Earthfile` is a
+lot like a `Dockerfile`, but with Make-like targets:
+
+```Dockerfile
+VERSION 0.8
+FROM golang:1.22
+WORKDIR /go-workdir
+
+deps:
+    COPY go.mod go.sum ./
+    RUN go mod download
+    # Output these back in case go mod download changes them.
+    SAVE ARTIFACT go.mod AS LOCAL go.mod
+    SAVE ARTIFACT go.sum AS LOCAL go.sum
+
+build:
+    FROM +deps
+    COPY main.go .
+    RUN go build -o output/example main.go
+    SAVE ARTIFACT output/example AS LOCAL local-output/go-example
+
+docker:
+    COPY +build/example .
+    ENTRYPOINT ["/go-workdir/example"]
+    SAVE IMAGE go-example:latest
+```
+
+You'd run `earthly +docker` to build the Docker target in this example.
+
+At first glance Earthly looks very similar to a multi-stage Dockerfile. There's
+a lot of overlap, but Earthly has a bunch of extra functionality that's useful
+for a general purpose build tool, including:
+
+* Invoking other Dockerized things ([`WITH DOCKER`][earthfile-with-docker]) -
+  e.g. Crossplane's E2E tests
+* Exporting files that changed in the build
+  ([`SAVE ARTIFACT AS LOCAL`][earthfile-save-artifact])
+* Targets that are simply aliases for a bunch of other targets.
+* The ability to import Earthfiles from other repos without a submodule
+  ([`IMPORT`][earthfile-import]).
+
+I feel Earthly's key strength is its Dockerfile-like syntax. Before writing this
+one-pager I ported 90% of Crossplane's build from Make to Earthly. I found it
+much easier to pick up and iterate on than the build submodule.
+
+### Performance
+
+Earthly is as fast as Make when developing locally, but a little slower in CI.
+CI is slower because the Go build cache doesn't persist across CI runs.
+
+Here are a few local development comparisons using a Linux VM with 6 Apple M1
+Max vCPUs and 20GiB of memory.
+
+| Task | Make | Earthly |
+| --- | --- | --- |
+| Build with a cold cache | ~46 seconds | ~60 seconds |
+| Build with a hot cache (no changes) | ~2 seconds | ~1 second |
+| Build with a hot cache (one Go file changed) | ~8 seconds | ~8 seconds |
+| Build for all platforms with a cold cache | ~4 minutes 10 seconds | ~4 minutes 40 seconds |
+| Build for all platforms with a hot cache (one Go file changed) | ~42 seconds | ~32 seconds |
+
+Here are some CI comparisons run on GitHub Actions standard workers.
+
+| Task | Make | Earthly |
+| --- | --- | --- |
+| Run linters | ~3 minutes | ~4 minutes |
+| Run unit tests | ~3 minutes | ~2.5 minutes |
+| Publish artifacts | ~12 minutes | ~14 minutes |
+| Run E2E tests | ~12 minutes | ~14 minutes |
```

Comment:
Curious @negz if you got a sense for any impact on reliability and/or performance of running e2e tests locally.  I just thought of this now while running `earthly -P +e2e` (which I believe runs `-test-suite=base`). It took about 20 mins to run that suite and it had 10 failures:

```
 DONE 114 tests, 13 skipped, 10 failures in 1196.054s
```

I'm not sure if that's slower or less reliable - have you developed a sense for that while testing yourself? 🤔

Reply:
No idea.

That said, I can't see why it would be slower. At the end of the day it's the same code running in a Docker container spun up using `kind` whether that was done by Make or Earthly.

Reply:
FWIW I just ran `earthly -P +e2e` locally and everything passed. I'm not sure how long it took though - I don't get the `DONE` summary for whatever reason.